Solutions/Corvus.Extensions/Corvus/Extensions/AsyncLinqExtensions.cs

```diff
@@ -0,0 +1,71 @@
+﻿// <copyright file="AsyncLinqExtensions.cs" company="Endjin Limited">
+// Copyright (c) Endjin Limited. All rights reserved.
+// </copyright>
+
+namespace CareerCanvas.Specs.Steps
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Text;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Asynchronous LINQ extensions.
+    /// </summary>
+    public static class AsyncLinqExtensions
+    {
+        /// <summary>
+        ///  An asynchronous version of the SelectMany operator.
+        /// </summary>
+        /// <typeparam name="T1">The type of the items in the source list.</typeparam>
+        /// <typeparam name="T2">The type of the items in the resulting list.</typeparam>
+        /// <param name="source">The source.</param>
+        /// <param name="mapper">The function used to map from each item in the source list to the result.</param>
+        /// <returns>A fanned out list of T2.</returns>
+        public static async Task<IEnumerable<T2>> SelectManyAsync<T1, T2>(this IEnumerable<T1> source, Func<T1, Task<IEnumerable<T2>>> mapper)
+        {
+            if (source is null)
+            {
+                throw new ArgumentNullException(nameof(source));
+            }
+
+            if (mapper is null)
+            {
+                throw new ArgumentNullException(nameof(mapper));
+            }
+
+            var result = new List<T2>();
+
+            return (await Task.WhenAll(
+                source.Select(sourceItem => mapper(sourceItem))).ConfigureAwait(false))
+                .SelectMany(result => result);
+        }
+
+        /// <summary>
+        /// An asynchronous version of the Select operator..
+        /// </summary>
+        /// <typeparam name="T1">The type of items in the source list.</typeparam>
+        /// <typeparam name="T2">The type of the items in the result list.</typeparam>
+        /// <param name="source">The source list.</param>
+        /// <param name="mapper">The function used to map to the results.</param>
+        /// <returns>A list of T2.</returns>
+        public static async Task<IEnumerable<T2>> SelectAsync<T1, T2>(this IEnumerable<T1> source, Func<T1, Task<T2>> mapper)
```

Comment:
As with the previous method, this doesn't feel very LINQy to me.

The essential character of `Select` in LINQ is that it doesn't change the nature of the container. E.g., `IEnumerable<T>` is designed for sequential lazy evaluation, and if you pass in an infinite source, `Select` returns an infinite source, projecting each item on demand. And just as the LINQ to Objects `Select` preserves this baisic `IEnumerable<T>` characteristic that items are produced one at a time as and when the consumer asks for them, the Rx `Select` preserves the basic `IObservable<T>` characteristic that items are produced one at a time as and when the producer decides to produce them.

It would be weird if an Rx `Select` operator decide to wait until it thought the source had finished before producing any items. More generally it feels weird for a `Select` operator to change the rules of who produces what and when.

So that's why I don't think this is `Select`. As with the previous method this feels like a handy wrapper around `Task.WhenAll`, so I don't think either of these should have LINQ-like names, or be in a class that mentions LINQ.

Also, my preference would be for them not to be extension methods. Not only does that make them align better with the `Task.WhenAll` method they resemble, it also means I don't get this rather async-only behaviour offered on absolutely every `IEnumerable<T>` just because I happened to bring in `using Corvus.Extensions;`. Or if you really want them as extension methods, maybe put them in a more specialized namespace.